usdt-impl/src/stapsdt.rs

```diff
@@ -0,0 +1,241 @@
+        .map(|provider| {
+            let provider = Provider::from(provider);
+            // Ensure that the name of the module in the config is set, either by the caller or
+            // defaulting to the provider name.
+            let config = crate::CompileProvidersConfig {
+                provider: Some(provider.name.clone()),
+                probe_format: config.probe_format.clone(),
+                module: match &config.module {
+                    None => Some(provider.name.clone()),
+                    other => other.clone(),
+                },
+            };
+            compile_provider(&provider, &config)
+        })
+        .collect::<Vec<_>>();
+    Ok(quote! {
+        #(#providers)*
+    })
+}
+
+pub fn compile_provider_from_definition(
+    provider: &Provider,
+    config: &crate::CompileProvidersConfig,
+) -> TokenStream {
+    compile_provider(provider, config)
+}
+
+fn compile_provider(provider: &Provider, config: &crate::CompileProvidersConfig) -> TokenStream {
+    let probe_impls = provider
+        .probes
+        .iter()
+        .map(|probe| compile_probe(provider, probe, config))
+        .collect::<Vec<_>>();
+    let module = config.module_ident();
+    quote! {
+        pub(crate) mod #module {
+            #(#probe_impls)*
+        }
+    }
+}
+
+/// ## Emit a SystemTap probe (version 3 format).
+///
+/// Source: https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation
+///
+/// A STAPSDT probe expands to a single `nop` in the generated code (the `nop`
+/// instruction is generated in `compile_probe`) and a non-allocated ELF note.
+/// Additionally, a special `.stapsdt.base` section is needed (once only) for
+/// detecting prelink address adjustments (its contents do not matter at all).
+///
+/// This method generates the ELF note assembly and an `.ifndef _.stapsdt.base`
+/// section for the address adjustments. Additionally, this method generates a
+/// 16 bit "semaphore" (counter) and links it to the ELF note. This semaphore
+/// is then used to gate invocations of the probe by reading its value at
+/// runtime and checking it against 0. A value of 0 means that no consumers are
+/// currently active and the probe and any parameter massaging work can be
+/// skipped.
+///
+/// ### Summary
+///
+/// A STAPSDT probe in plain pseudo-Rust would look roughly like this:
+/// ```rs
+/// #[linkage = "weak", visibility = "hidden"]
+/// static __usdt_sema_provider_probe: SyncUnsafeCell<u16> = SyncUnsafeCell::new(0);
+///
+/// let is_enabled = unsafe { __usdt_sema_provider_probe.get().read_volatile() } > 0;
+///
+/// if is_enabled {
+///   let arg1: u64 = get_arg1();
+///   let arg2: u32 = get_arg2();
+///   let arg3: *const u8 = get_arg3();
+///   const {
+///     // Build time: Define this location as being the place in the
+///     // instruction stream where our probe is located.
+///     core::usdt::define::<(u64, u32, *const u8)>("provider", "probe", &__usdt_sema_provider_probe);
```

Comment:
The `core::` prefix is a bit misleading. This is a stand-in for the inline assembly below, right? If so, I might remove that.

Reply:
Removed the `core::` prefix and added a bit more comments around here to hopefully make the intent of this pseudo-code a little bit clearer yet.